### DISCUSSION TITLE
Hashing the invoice, specs

### DISCUSSION THREAD

**Discussion by ec-ecss** [mrsool/zatca#26] [General]
Hello 
Could you provide (if it's possible) an exemple of loading a existing generated xml invoice file (supposed to be valid )
And then generate the Zatca hash ?

**Comment by obahareth** [comment]
Hey @ec-ecss,

This functionality is planned but not yet available.

The idea was to make a `from_xml` on each XML element to eventually read it back to a `Invoice` object.

We didn't have a real use case for this though. If you need it quickly you could use ZATCA's SDK (check the einvoicing-sdk folder, we dockerized it for ease of use)

`fatoora -generateHash /path/to/invoice.xml`

Could you elaborate more on what your use case is?

**Comment by ec-ecss** [comment]
I tried to use the sdk hash command, but it seems that the sdk hash resulted in something that was not recognized by the API...
While your example code was directly working.
And it's the only code in all the web I found that was working 
I finaly manage to write a ruby script that is based on your package, that take one invoice as argument and generate b64 and hash.
Are you interested ?

**Reply by obahareth** [reply]
@ec-ecss Yeah sure! Would be great if you could share it, might speed up the process of adding this functionality in the SDK.

**Reply by ec-ecss** [reply]
Here is my code !
https://github.com/mrsool/zatca/pull/28
There is repetition for the "hack sign template" but I'm a Ruby newbie and I don't know how to use the existing code in a static way.
I hope it will help

**Comment by ec-ecss** [comment]
The use case is simply I've already got my code with xml ubl working generation that check every shema and shematron to qualify ( that is based on my client workflow )
But get the api to work and upload to zatca is a serious pain
